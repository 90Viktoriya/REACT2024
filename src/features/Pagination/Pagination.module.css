.pagination {
  width: 100%;
  justify-content: center;
  display: flex;
  flex-direction: row;
  min-height: 40px;
}
.inactive {
  border: 2px outset var(--dark);
  pointer-events: none;
  cursor: none;
}
.active {
  border: 2px outset var(--dark);
  cursor: pointer;
}