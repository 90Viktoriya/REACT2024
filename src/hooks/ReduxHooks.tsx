import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import type { RootState, AppDispatch } from '../store/storeRedux';

export const useAppDispatch = useDispatch.withTypes<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
//export const useAppSelector: RootState = useSelector.withTypes<RootState>();
