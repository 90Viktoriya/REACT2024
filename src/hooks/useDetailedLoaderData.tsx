import { useLoaderData } from 'react-router-dom';
import { detailedLoader } from '../features/Router/Loaders/detailedLoader';

export const useDetailedLoaderData = () => {
  return useLoaderData() as Awaited<ReturnType<typeof detailedLoader>>;
};
