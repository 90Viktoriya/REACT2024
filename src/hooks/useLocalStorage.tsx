import { useState, useEffect } from 'react';
import { Store } from '../store/store';

export function useLocalStorage() {
  const [searchValue, setSearchValue] = useState<string>(() => Store.getSearchValue());
  useEffect(() => {
    if (searchValue === undefined) {
      return;
    }
    Store.setSearchValue(searchValue);
  }, [searchValue]);
  return { searchValue, setSearchValue };
}
