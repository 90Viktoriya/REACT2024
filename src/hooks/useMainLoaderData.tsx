import { useLoaderData } from 'react-router-dom';
import { mainLoader } from '../features/Router/Loaders/mainLoader';

export const useMainLoaderData = () => {
  return useLoaderData() as Awaited<ReturnType<typeof mainLoader>>;
};
