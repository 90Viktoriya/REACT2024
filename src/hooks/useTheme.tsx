import { useContext } from 'react';
import { ThemeContext } from '../features/Theme/ThemeContext/ThemeContext';

export const useTheme = () => useContext(ThemeContext);
