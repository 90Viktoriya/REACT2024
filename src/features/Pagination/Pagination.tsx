import { Link } from 'react-router-dom';
import styles from './Pagination.module.css';
import { ComponentsCaptions } from '../../data/ComponentsCaptions';
import { RouterPath } from '../Router/Router.enum';
import { useAppSelector } from '../../hooks/ReduxHooks';
import { useGetCharactersByNameQuery } from '../../services/apiRTK';

export function Pagination() {
  const searchValue = useAppSelector((state) => state.navigation.searchValue);
  const currentPage = useAppSelector((state) => state.navigation.currentPage);
  const { data } = useGetCharactersByNameQuery({ name: searchValue, page: currentPage });
  const page = data?.page;

  return (
    <div className={styles.pagination}>
      <Link
        to={currentPage > 1 ? `/${RouterPath.PAGE}/${currentPage - 1}` : `../`}
        className={page?.firstPage ? styles.inactive : styles.active}
      >
        {ComponentsCaptions.PREV}
      </Link>
      <Link
        to={`../${RouterPath.PAGE}/${currentPage + 1}`}
        className={page?.lastPage ? styles.inactive : styles.active}
      >
        {ComponentsCaptions.NEXT}
      </Link>
    </div>
  );
}
