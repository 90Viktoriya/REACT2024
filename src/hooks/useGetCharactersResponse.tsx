import { useGetCharactersByNameQuery } from '../services/apiRTK';
import { useAppSelector } from './ReduxHooks';

export function useGetCharactersResponse() {
  const searchValue = useAppSelector((state) => state.navigation.searchValue);
  const currentPage = useAppSelector((state) => state.navigation.currentPage);
  const { data } = useGetCharactersByNameQuery({ name: searchValue, page: currentPage });
  return data;
}
