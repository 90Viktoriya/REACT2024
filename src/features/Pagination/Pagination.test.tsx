import { render } from '@testing-library/react';
import { Pagination } from './Pagination';
import { Mock, vi } from 'vitest';
import { MemoryRouter } from 'react-router-dom';
import { useGetCharactersResponse } from '../../hooks/useGetCharactersResponse';

vi.mock('../../hooks/useGetCharactersResponse', () => ({
  useGetCharactersResponse: vi.fn()
}));
vi.mock('../../hooks/ReduxHooks', () => ({
  useAppSelector: vi.fn()
}));

describe('Given Pagination component', () => {
  it('when rendered, should match snapshot', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <Pagination />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });
  it('when first page, prev button should be inactive', () => {
    (useGetCharactersResponse as Mock).mockReturnValue({ page: { firstPage: true, lastPage: false } });
    const screen = render(
      <MemoryRouter>
        <Pagination />
      </MemoryRouter>
    );
    const link = screen.queryByText('Prev');

    expect(link?.className).toMatch(/_inactive/);
    expect(link?.className).not.toMatch(/_active/);
  });
  it('when not first page, prev button should be active', () => {
    (useGetCharactersResponse as Mock).mockReturnValue({ page: { firstPage: false, lastPage: false } });
    const screen = render(
      <MemoryRouter>
        <Pagination />
      </MemoryRouter>
    );
    const link = screen.queryByText('Prev');

    expect(link?.className).not.toMatch(/_inactive/);
    expect(link?.className).toMatch(/_active/);
  });
  it('when last page, next button should be inactive', () => {
    (useGetCharactersResponse as Mock).mockReturnValue({ page: { firstPage: false, lastPage: true } });
    const screen = render(
      <MemoryRouter>
        <Pagination />
      </MemoryRouter>
    );
    const link = screen.queryByText('Next');

    expect(link?.className).toMatch(/_inactive/);
    expect(link?.className).not.toMatch(/_active/);
  });
  it('when not last page, next button should be active', () => {
    (useGetCharactersResponse as Mock).mockReturnValue({ page: { firstPage: false, lastPage: false } });
    const screen = render(
      <MemoryRouter>
        <Pagination />
      </MemoryRouter>
    );
    const link = screen.queryByText('Next');

    expect(link?.className).not.toMatch(/_inactive/);
    expect(link?.className).toMatch(/_active/);
  });
});
