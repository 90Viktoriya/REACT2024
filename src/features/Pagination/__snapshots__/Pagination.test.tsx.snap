// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Given Pagination component > when rendered, should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="_pagination_f56b67"
  >
    <a
      class="_active_f56b67"
      href="/"
    >
      Prev
    </a>
    <a
      class="_active_f56b67"
      href="/page/NaN"
    >
      Next
    </a>
  </div>
</DocumentFragment>
`;
